docs/adr/000x-modelling-history.md

```diff
@@ -0,0 +1,250 @@
+significantly Revision and its associated types. These models should be
+persisted to the database once and never updated or deleted, any need to change
+them requires creating a new record. This allows us to store a full history
+by only appending to the database.
+
+Accessing immutable models is intended to always be done by a foreign key and
+not by the usage of `SELECT MAX` style queries. This is done for
+simplicity, performance and for consistency with Rails idioms. To achieve this
+there are a some mutable models, eg Edition, that reference immutable models,
+eg Revision. When the content for an edition is revised the Edition model
+references the new Revision. Mutable models should not be used to store fields
+where there is a need to know how the value of that field changed over time.
+
+The choice of this immutability strategy is to store both present and
+historical concerns in the same data stores, thus ensuring history remains a
+first class citizen. A nice side effect of having immutable models is
+this opens options for caching. Since data for that
+model will never change it can effectively be cached forever.
+
+### Breakdown of Revision
+
+Revision is a model that stores little data itself and joins to other models
+that store the data. This can be visualised as:
+
+![Revision breakdown](000x/revision-diagram.png)
+
+The intention of breaking this up is to be conservative with the amount of data
+duplicated for both storage and maintenance reasons. Thus when a user edits
+the title of an edition a new ContentRevision object is created and the existing
+objects (TagsRevision, MetadataRevision and ImageRevisions) are associated with
+the next revision. An ImageRevision is modelled in a similar way to a Revision
+and this is explained further in [Image modelling](#image-modelling).
+
+It is intended that [delegation][delegate] be used when interfacing with a
+revision so that the caller need not be concerned with which sub-revision
+stores particular fields. This allows a revision to have a rich interface
+despite storing a low amount of data directly.
+
+### Image modelling
+
+Content Publisher supports a user uploading image files and referencing them
+in a revision of a document. They have metadata and editable properties that a
+user can change, of which a history is stored. A single image file uploaded
+produces multiple files that are uploaded to Asset Manager for different sizing
+variations. Images are modelled in a similar way to Revision with an
+Image::Revision model, as represented below:
+
+![Image Revision breakdown](000x/image-revision-diagram.png)
+
+The Image model itself is used for continuation between image revisions. It is
+known that two Image::Revisions are versions of the same item if they share the
+same Image association.
+
+The data of an Image::Revision is stored between an Image::FileRevision and an
+Image::MetadataRevision. These two objects differ by the fact that any change to
```

Comment:
'two objects' -> 'two models'

Reply:
Could we rephrase to say the less important thing first? Something like 'Both have the same immutable behaviour as the sub-revisions of a Revision, but any change to...'

Reply:
Could we use 'assets' instead of 'Asset Manager files', as the latter doesn't seem to be defined - it's at least unclear whether it's a concept in our app or external (or both ;-).

Reply:
I've changed this around a bit.

As per other comment, you have a point but I think "asset" requires more qualifying and comparing the terms in context we mention "files that are uploaded to Asset Manager" in the previous sentence so this terminology follows.